### ISSUE TITLE
Render ref from route relationship

### ISSUE THREAD

**Issue by matthijsmelissen** [gravitystorm/openstreetmap-carto#596]
The following issue has been moved over from [trac](https://trac.openstreetmap.org/ticket/2610). It needs more discussion as it is not clear how this should be implemented.

Highway shields are only rendered when the ways themselves have a ref tag. Now, the different ways that form a road route are assembled into relations. For the sake of avoiding duplicate data the ref tags should be removed from the individual ways and applied to the relation only.

However, currently we only use ref tags on roads for rendering, and we ignore the ref tag on the relation.

Any idea how we could implement this?

**Comment by pnorman** [comment]
Unassigning because I have no plans of working on this in the near or medium term, and I don't want others to not work on it because they think I am.

I will caution for anyone planning on picking this up that this has been a fairly long-running issue, reading it is somewhat difficult because

- a large number of people suggesting "easy" mechanisms which are not technically possible, so you need to know what will work
- sections are off-topic, talking about pictorial shields
- people assume that their local tagging practices apply world-wide

If you want to take a stab at implementing it, the requirements are
- it works without slim tables
- route relations being [split by osm2pgsql differently than ways](https://github.com/gravitystorm/openstreetmap-carto/issues/596#issuecomment-47410719) is handled
- the tagging practice of ref information on ways **and** the practice of ref information on route relations are **both** handled
  - ref information simultaneously on ways and relations is an open question
- When writing the query, performance and readability need to be considered

**Comment by SomeoneElseOSM** [comment]
(just spotted this while looking at https://github.com/gravitystorm/openstreetmap-carto/issues/508#issuecomment-328315877 )

The comments above (admittedly some of which date from 2014) seem to assume an entirely SQL-based approach.  An alternative and frankly much less horrible option would be process the route relations in lua first (as per http://www.openstreetmap.org/user/SomeoneElse/diary/38988 ).  This could control as much or as little in the map display as needed (perhaps only shields).

**Comment by pnorman** [comment]
> The comments above (admittedly some of which date from 2014) seem to assume an entirely SQL-based approach. An alternative and frankly much less horrible option would be process the route relations in lua first (as per http://www.openstreetmap.org/user/SomeoneElse/diary/38988 ). This could control as much or as little in the map display as needed (perhaps only shields).

The problem is because we need to handle shields from either relations or the ways, **and** we need to colour shields based on **way** information, not relation information we need to deal with some complexities that don't apply to what you're doing.

**Comment by skylerbunny** [comment]
> The problem is because we need to handle shields from either relations or the ways

I don't think there's a need to do that at all. In fact, there's a very good argument to only generate pictoral shields (meaning, images likely based on svg vectors) based on relations. My thought for why is, that we can take advantage of tags that route relations get, that ways don't - particularly, ```network```:

* If at least one relation on a way exists with a route type and a ```network=``` key, render one or more sprites based on that network key, possibly leveraging a ```symbol``` svg tag.

* If not, fall back to what we do now, which is simple way tags based on a ```ref=``` key.

**Comment by skylerbunny** [comment]
(Practical implementation example of the above: https://wiki.openstreetmap.org/wiki/Wisconsin/State_Highway_Relations )

I'd have no expectation that state highways, county highways, or other to have vector shields generated if they are NOT part of a relation of this sort. The computation intelligence to determine if a way 'without a referenced relation' should display a pictoral shield would be hell:

https://www.openstreetmap.org/way/111537836 : "NC 15" - is it North Carolina State Highway 15? No, because it's in Africa. This implies we have to evaluate location context for every way to make sure that a ref string makes sense for the area in question. This leads to...

If a way shares two route designations, how do we define those for a way? Semicolon separated network lists which should be in the same order as a semicolon separated reference list?

I could think of more examples, but I really think limiting this to discrete relations like those Wisconsin has prevents a lot of heartache.

**Comment by skylerbunny** [comment]
...ugh. Well, I see that this is only dealing with relations in a list, not shields at all.

Apologies.

I'll modify my comments appropriately. But in those cases I still think it would be best to let relations override way tagging entirely if it exists on a way. The principle is the same.

**Comment by pnorman** [comment]
> I'll modify my comments appropriately. But in those cases I still think it would be best to let relations supercede way tagging if it exists on a way, only adding way tags if they're different, at the end. The principle is the same.

This doesn't change the problems of complexity - it's coloring the shields based on the highway=* tag, not dealing with ref information on both.

**Comment by 1ec5** [comment]
> If you want to take a stab at implementing it, the requirements are
> 
> * it works without slim tables
> * route relations being [split by osm2pgsql differently than ways](https://github.com/gravitystorm/openstreetmap-carto/issues/596#issuecomment-47410719) is handled
> * the tagging practice of ref information on ways **and** the practice of ref information on route relations are **both** handled
>   * ref information simultaneously on ways and relations is an open question
> * When writing the query, performance and readability need to be considered

@kennykb, to what extent does the new [osm-shields](https://github.com/kennykb/osm-shields/) project satisfy these requirements?

**Comment by SomeoneElseOSM** [comment]
> Solving https://github.com/openstreetmap/osm2pgsql/issues/230 seem by far the best approach to this.

The key bit that needs an osm2pgsql change is if you want to render route refs differently (e.g. in a different colour) depending on the type of highway way it's being displayed on.  If a US-centric map was happy to use the same style of ref representation ("shield" in US terms I guess) regardless of highway type then I don't believe you'd need an osm2pgsql change.  As an example, the walking and cycling routes at https://map.atownsend.org.uk/maps/map/map.html#zoom=14&lat=54.12077&lon=-0.96084 are done by directly rendering from the relation.

**Comment by imagico** [comment]
Yes, but if you want to arrange multiple shields in a certain way it is still fairly awkward to do so based on the current _individual ways plus separate route geometries_ structure generated by osm2pgsql.  Hence my general remark regarding the best approach.

**Comment by kennykb** [comment]
Please note that I did my osm-shields for my own use - the proposal to adopt it for the main rendering chain is not mine! I shared it primarily because I could not find a worked example of how to use the GroupSymbolizer to render marker clusters for concurrent routes in multiple networks, and encountered enough difficulty trying that a project to share the results seemed to be in order.

With that said, don't be quick to discount the possibility that this approach could be reworked in such a way as to scale to the main map.

**minutely updates**

The current implementation works entirely without either parsing the OSM changeset or modifying osm2pgsql. It's actually pretty well suited to the 'daily update from GeoFabrik' workflow. Minutely updates are a different workflow. 'Incompatible with minutely updates,' as @imagico puts it, is _far_ too strong a statement; 'minutely updates need a little more coding, ideally with some help from osm2pgsql' is more accurate. To me, it's quite obvious how to handle them, and I give three possible integration strategies on the [project wiki](https://github.com/kennykb/osm-shields/wiki/Maintaining-the-association-between-ways-and-routes#incremental-update-to-the-tables). The integration alteratives depend more on social issues than technical ones; if osm2pgsql cannot be modified to maintain the auxiliary tables, then fall back on reparsing the changes delivered by osmosis, or use database triggers to keep them in sync.

I haven't troubled to develop the minutely workflow, because I want to reassure myself that it won't be a complete waste of time. Given that I'm not all that well connected in the community, I strongly suspect that I may not be able to negotiate the social and political hurdles of developing to the 'big' rendering systems, or that the approach will fail to scale in some way that I have not anticipated. I haven't the resources personally to host a full-planet, minutely-updating system, so testing would be a problem. The technical problem is easy compared with getting access to the facilities needed to test a solution, and getting such a solution integrated into the standard tools. The latter issues are what worry me.

With regards to the other remarks from @pnorman, relayed by @1ec5:

**slim tables** 

@imagico statement that the implementation 'relies on slim tables' is also rather too strong a statement. It relies on the slim tables to perform its updates - _just as osm2pgsql does._ The slim tables are not used at all at render time.

What is used at render time is three new tables. One has a row for each route=road relation, and maintains the 'network' and 'ref' tags. A second has a row for each way that is a member of one or more route=road relations, and holds simply the relation ID, the way ID, and the role. The third is a directory that maps network/ref pairs to the file names for the corresponding graphics.

Having something in the database to identify where to find the graphic elements is essential. Alas, US road networks are anything but a 'small fixed set' and so need a database to track them. The other two tables could, I suppose, be replaced with a modification to osm2pgsql that would copy the network and ref information down to the individual items in the line table. This would, I suspect, prove less performant (because of replicated data) than the JOIN operation needed to go from 'I am rendering this set of ways' to 'this is the set of route markers that I need to place'. Given that a schema change to support the graphic elements would be required anyway, I decided to go with the JOIN.

Having the tables available for joining allows for queries to go in both directions: 'what is the set of ways that I will encounter if I follow County Road 34?' and 'what is the set of route markers that I may expect on way 2718281828?'

In order to have the smallest possible footprint, I confine myself to relations with 'type=route', 'route=road', and 'network' and 'ref' both present. Obviously,

**approach to splitting linestrings**

The osm2pgsql approach of reassembling route relations back into long linestrings is fundamentally not going to work with route concurrencies, and in fact I totally ignore it and work with the original ways plus the auxiliary tables. The problem with the approach is that the linestrings need to represent distinct _sets_ of relations, not individual relations. In the diagram below, the red and blue lines represent distinct routes. In the section where the routes are concurrent, there's really no practicable way for a renderer to tell that if all it has is the separate red and blue lines. It really needs five linestrings (or three multilinestrings) in this case: the bits that belong only to the red route (two linestrings or one multilinestring), the bits that belong only to the blue route (again, two linestrings or one multilinestring), and the segment that is shared by both (one linestring).

![concurrency](https://user-images.githubusercontent.com/216913/44659534-8c7c9000-a9d2-11e8-8039-87ef31137b47.png)

The implementation as coded defers splitting and merging linestrings to render time, when there will always be a set of manageable size (because it consists only of the ways that will be rendered on one supertile).  It uses the powerful PostGIS 'ST_LineMerge' operation to group ways that belong to the same _sets_ of routes together and coalesce them where possible. About the only thing that it runs afoul of is that Mapnik winds up trying to place a marker on each carriageway of a motorway, even at low zoom levels, leading to a proliferation of markers. This appears to be a deviation from what the Mapnik documentation promises, and I've entered a query at  [mapnik/mapnik/issues/3978](https://github.com/mapnik/mapnik/issues/3978) about how best to solve the problem.


** handling both the tagging practice of putting the ref on the way and putting it in the relation **

Since I was focusing on a North American map, and the practice of placing route references on relations has become near-universal here, I simply didn't bother coding up the alternative. As long as you don't need to have both styles of tagging supported _on the same way_, it's easy to implement. Existing tagging practice in the US will mean that the ways with the tag on the way will not get shaped markers, but rather use something like today's OSM style. It's too long a leap to expect the renderer to take a ref that looks like 'CR 34', figure out from administrative boundaries what county the 'CR' refers to, and look up the appropriate marker.

** performance and readability **

The performance appears adequate. On the one largish (about a third of the United States) map that I render, the rendering remains I/O bound - it's dominated by the fact that on my machine, owing to fairly limited SSD space, rendering a supertile takes at least two seeks on magnetic disks, one to retrieve a hillshading image and another to retrieve raster-encoded land cover data.

It is, of course, impossible to offer assurances regarding performance in the fielded environment without testing in a very similar setup to the production system. I do believe that what performance problems we will encounter are surmountable.

As far as readability goes, given that the assembly of route concurrences is done in the database, we are dealing with some fairly complex SQL here.  I can, of course, comment it much more extensively, but there is complexity there that simply cannot be eliminated while continuing to satisfy all the other constraints.  From Mapnik's point of view, the complexity is reduced to a simple call on a stored procedure. Within the stored procedure, [this wiki page](https://github.com/kennykb/osm-shields/wiki/Using-the-GroupSymbolizer) provides an overview of what is happening. Of course, I plan to put that discussion in abbreviated form in the commentary in the stored procedures themselves.

**Comment by imagico** [comment]
I think you misunderstood my statements - i did not mean to pass judgement on the suitability of your approach for what you are trying to accomplish (for that i have not looked even remotely close enough at your code).  I just tried to answer the question by @1ec5 regarding the suitability of your code for this style based on the requirements previously stated in this issue.

And i disagree that _relies on slim tables and is incompatible with minutely updates_ is too strong, it is an accurate description of the current situation as far as i can see.  If your code can be modified to allow minutely updates in a way that works reasonably well on a world wide database i don't know.

**Comment by kennykb** [comment]
@imagico: Is there an approach to minutely updates in which osm2pgsql does not rely on slim tables? Can you give me a link to a discussion of such a thing? I see on the OSM wiki:

> Some important features (including incremental updates (planet diffs, the initial load to populate the track tables, and proper evaluation of multipolygons) only work in slim mode.

There's very likely something of which I'm unaware here. Otherwise, it seems rather odd that osm2pgsql may depend on slim tables to perform incremental updates and apply planet diffs to 'roads', 'line', 'point' and 'polygon', but I may not depend on them for incremental update of the added tables that support route concurrences. That's why I was careful to clarify that I do _not_ depend on the slim tables at render time, only to perform updates.

**Comment by kennykb** [comment]
I suppose I should make a progress report.  The code in osm-shields now handles ref=* tags on either a route relation or directly on a way. (The ones on ways are intentionally not processed if the way participates in a group relation.)

See [the corner of New Hampshire, Vermont and Massachusetts](https://kbk.is-a-geek.net/catskills/test4.html?la=42.7319&lo=-72.4610&z=11) for a transition from "refs mostly on route relations rendered with pictorial shields" to "refs mostly on the way itself, rendered in classic OSM style"

Concurrencies among route relations are handled with the GroupSymbolizer.
Multiple ref=* values on a single way are handled by stacking them in the box as seen [near Kingston, NH](https://kbk.is-a-geek.net/catskills/test4.html?la=42.9363&lo=-71.0517&z=13).

Let this serve as a proof of concept that having both representations in the same map isn't a show-stopper.

**Comment by skylerbunny** [comment]
If it helps to do some spot checking, I did some fixes in the last couple hours to the state of Georgia, at `US:US`, `US:GA`, and their sublevels like counties and modifiers, to try to make them consistent; that is, modifiers exist where applicable, and networks reflect 'a county or the same modifier that appears in the modifier tag' at the third network level. Some of them were out of sync with this pattern, had `GA` or `US` or a modifier in the ref, and so forth.

Georgia seems to be pretty mature in terms of 'most routes at the state and US level have relations, and counties have a few too.' I wouldn't claim I found every error in the above, but I do think I got a lot of them. I found a couple dozen.

**Comment by pnorman** [comment]
> If you want to take a stab at implementing it, the requirements are

In addition to the requirements I listed, there are the global ones for the stylesheet
- That it is written in CartoCSS
- That it doesn't use views, extra tables, or anything which is not created by the current install instructions
- That it meets our code style guidelines

To solve this in osm2pgsql, implement #230. This is a big job because it in turn requires changes to osm2pgsql's Lua API.

-----

It is possible to implement this meeting all of the criteria except SQL readability and performance. I [did so in 2014](https://github.com/gravitystorm/openstreetmap-carto/issues/596#issuecomment-57001666) but the SQL was too messy for **me**.

If I were to tackle it today I'd start by cleaning up the roads SQL. If that were cleaner, it'd be easier to work with, and new complexity would be more tolerable. I'm not sure there's a direct connection on the SQL side, but cutting back the roads MSS would certainly help.

Then comes the tough question of how to implement. The requirements of working with osm2pgsql as we call it (no slim tables) and understandable SQL are hard requirements. I'd see if one of the requirements about features needed could be weakened.

If none of them can, then I don't believe it's possible to solve, given the constraints.

**Comment by kennykb** [comment]
Given the constraint that you express that 'the schema as implemented with a minimalistic osm2pgsql import today cannot be expanded in any way in the future', I think you are right that the problem cannot be solved.

If you are indeed that adamant, I would also assume that any change to osm2pgsql to create auxiliary tables to track membership in route relations would also be summarily rejected. (I further assume that this would be the case even if enabled only on request by a command-line option.)

The schema limitations are the one constraint that I find hobbles me. I can live with all the others. Well, maybe not with the lack of GroupSymbolizer, but I think I could duplicate its function in the pipeline that generates graphics, and then use prerendered groups in Mapnik. It would probably be easier to implement support for GroupSymbolizer in CartoCSS, but I'm presuming that an attempt at _that_ would also be rejected out of hand.

If I'm correct that you are unalterably opposed to the ideas, I'd rather find out sooner or later, since I don't want to find myself in the position of maintaining a fork. I'd rather spend the effort that I'd put into osm2pgsql on retooling to some other piece of software.

If you're still reading this, I quite agree with you about the 'roads' table. I understand that compatibility will require retaining it, but I hardly use it. If I create indices using statements like:
```sql
CREATE INDEX idx_line_primary ON planet_osm_line USING GIST(way) WHERE highway IN ('motorway', 'trunk', 'primary')
```
I find that I hardly miss it. I deal with routes using the tables that track membership of the individual ways in route relations. 

By the way, those tables are tiny by OSM standards - for North America, there are about fifty thousand route=road comprising about a million and a half ways. If I were to replace the 'role' column with a simple flag indicating whether a way is or is not 'role=link', each row in the 1.5-million-row table would be 25 bytes (8 for osm_id of the relation, 8 for osm_id of the way, 4 for position of the way within the relation, 1 for flags). Joining through it is cheap, since PostgreSQL tends just to keep in in memory while rendering is going on. And it's surely not something that would add hours to an import.

**Comment by gravitystorm** [comment]
> Given the constraint that you express that 'the schema as implemented with a minimalistic osm2pgsql import today cannot be expanded in any way in the future', I think you are right that the problem cannot be solved.

I think there's some miscommunication here. Paul was saying that anything like schema changes or extra tables needs to be upstream in osm2pgsql, not done here in this style. ( His link should have been  to the relevant issue in the osm2pgsql repo, namely https://github.com/openstreetmap/osm2pgsql/issues/230 )

What he then explained was that it is in fact possible to solve the relation symbolizers without changing how osm2pgsql works, but that it leads to very messy SQL. So it's again highlighting that solving that osm2pgsql issue is the way forward. Please don't assume that fixes provided to osm2psql will be rejected.

> It would probably be easier to implement support for GroupSymbolizer in CartoCSS, but I'm presuming that an attempt at that would also be rejected out of hand.

I can say as a CartoCSS maintainer that GroupSymbolizer support would be welcomed.

> If I'm correct that you are unalterably opposed to the ideas, I'd rather find out sooner or later, since I don't want to find myself in the position of maintaining a fork. I'd rather spend the effort that I'd put into osm2pgsql on retooling to some other piece of software.

There are hundreds of downstream users of osm2pgsql that would appreciate better route relation handling! So please do consider discussing your ideas on that tracker, and I hope to see your efforts there.

**Comment by kennykb** [comment]
I will concede I got confused by the fact that @pnorman wears multiple hats - and misinterpreted him as speaking in his rôle as lead developer of osm2pgsql, not strictly as a developer of the OSM Carto style.

**Comment by kennykb** [comment]
Bolstered by the encouragement from @gravitystorm,  I'm floating a formal [proposal](https://github.com/kennykb/osm-shields/wiki/Proposal:-add-route-tables-to-osm2pgsql) to develop the necessary _osm2pgsql_ infrastructure to address the problem.

Consider this to be a request, nay, rather, a plea for comments from those who understand the issues. The proposal is most emphatically not in its final form - in particular, there is an [Open issues](https://github.com/kennykb/osm-shields/wiki/Proposal:-add-route-tables-to-osm2pgsql#open-issues) section that describes some requirements that I'm fairly sure exist, but that I don't understand.

I hope that writing up this portion of the issue, as I understand it, demonstrates a good-faith attempt to "do the homework" and will not be dismissed out of hand as yet another ignorant oversimplification of the problem. I recognize that I'm still at an early stage in understanding, but I'm reluctant to move forward without at least some feedback from wiser heads. I'll make every effort to insure that your help will not be a waste of time.

**Comment by jeisenbe** [comment]
The issue openstreetmap/osm2pgsql#230 is [now fixed](https://github.com/openstreetmap/osm2pgsql/issues/230#issuecomment-583742025) by  [osm2pgsql PR1037](https://github.com/openstreetmap/osm2pgsql/pull/1037), so it should be possible to implement this. 

Is anyone interested in submitting a PR to render the ref from the route relation in addition?

**Comment by SomeoneElseOSM** [comment]
> Is anyone interested in submitting a PR to render the ref from the route relation in addition?

Presumably it'd need more than a change to this map style - it'd also need the way that osm2pgsql is used on osm.org to be invoked differently?

**Comment by kennykb** [comment]
@jeisenbe I'm following the development of the 'flex' backend eagerly, but my time is severely limited, possibly for the entirety of 2021 (we'll see how things start sorting out later in the year.

I'm considering the rendering of route relations on the main server to be Out Of Scope for me at the present time, owing to the fact that performance is unproven. Paul and Sarah are rightly concerned that reconstruction of concurrencies at render time might be an unacceptable cost. I am surely not tooled to assess the performance at production scale, so without a champion on the team, I'd rather be blocked.

I will confess that I suspended further development of my own pictorial-shield rendering because the performance worries extended to an indication that the functionality, the way I use it, would never be workable using a stock 'osm2pgsql'. I had no desire to maintain a fork, and frankly lacked energy to assess a switch to ImpOSM or another database pipeline.

Could @jeisenbe or another developer of the 'flex' backend revisit [the schema I'm using](https://github.com/kennykb/osm-shields/wiki/Proposal:-add-route-tables-to-osm2pgsql#proposal-part-1-add-additional-tables-to-track-route-members) and verify that it would be feasible to

- populate the tables - as specified - on initial import
- update the tables when processing diffs
- generate some sort of log of diffs to drive tile expiry (I'm willing to accept this in any form that's convenient. Processing updates is less time-critical than rendering tiles; I can reconstruct the geometry to determine the affected tiles myself.)

The key structural differences are:

- `planet_osm_route` lacks a geometry. It has only an OSM ID and tags.
- `planet_osm_route_way` is a new table that has one row per relation membership.  (Ideally, we should cluster this one on `planet_osm_route`.) It, too, has no geometry. Unlike other tables in the older schema, its rows aren't tied to a single OSM ID; instead, it represents the association between relation and member.

I'm at a point where I critically need a second pair of eyeballs on most of the route rendering stuff. Paul and Sarah have indicated that they consider my current design unacceptable, but have offered only sparse guidance as to what requirements it fails to address or how to move it forward. (I don't expect anything else. They're busy individuals, and I'm not paying them for the work.) I have the sense that something is badly wrong with the ideas, but I've no clue what it is!

**Comment by jeisenbe** [comment]
We should check if release 1.4.0 of osm2pgsql provides the proper functionality: https://github.com/openstreetmap/osm2pgsql/releases/tag/1.4.0

**Comment by kennykb** [comment]
@jeisenbe - The `flex` backend, plus a little postprocessing script, works out nicely.

I have it processing minutely update on my home server (which has data for only North America, but isn't an incredibly powerful machine) for several months now and the data import appears to be stable.

I have three big issues to overcome.  (1) I refuse to stand up Apache on this machine and mod_tile appears to be the only tile caching engine that isn't abandonware. (2) @pnorman has indicated that any table join in the rendering pipeline is unacceptable, and I haven't yet found a way to denormalize the data to satisfy this, (3) Carto-CSS (not OSM-Carto, the underlying system) doesn't do the GroupSymbolizer for rendering shields.  These issues obviously cut across multiple big code bases and it's hard for me to get a handle on all of them at once. I'd say that `osm2pgsql` is no longer part of the problem.

**Comment by pnorman** [comment]
> @pnorman has indicated that any table join in the rendering pipeline is unacceptable, and I haven't yet found a way to denormalize the data to satisfy thi

Joins are okay. We already have them. Using osm2pgsql's internal tables isn't.

**Comment by kennykb** [comment]
@pnorman Glad to hear that.

I'm not using the slim tables at all - that's all replaced with `shieldroute` and `shieldway` tables created in the Flex backend at https://github.com/kennykb/osm-shields/blob/master/shieldtables.lua plus a script that runs after the update to make sure that there are appropriate SVG's available for the shields (I have SVG templates with up to four subsitituted values that make those, so as to make stacks like `Alternate` `New York` Route `9` `N` (the '9' and 'N' are in different fonts) 

The update procedure has been running minutely on my box, in front of a PostGIS database that started from a `northamerica` extract from GeoFabrik, since November, without a hiccup, so the 'minutely update' and 'slim tables' problems are 100% solved.

Actually assigning route to way is moderately complex (hadling concurrencies without breaking marker placement is kind of difficult), and it's done in a set of stored procedures.  They look to be hideous multiway JOIN operations, but in practice the common table expressions get optimizied and virtually all operations are in-memory. While the procedures are complicated, I think they're reasonably well documented.  https://github.com/kennykb/osm-shields/blob/master/queryprocs.sql.in

If you're now interested, the structure is that `query_shields_(line_all|roads_all|roads_major|motorway)' are the procs that the renderer calls at various zoom levels, and they all are thin wrappers around `analyze_markers`. That function breaks down into two parts. The first produces shaped shields for route relations. The second produces OSM-style markers for ways that have OSM-style refs but no corresponding route relations.

The pipeline for route relations also has a fallback case for OSM-style markers if there's no special graphic available for a given network.

There's a whole insane mess of Tcl code (which I'm entirely willing to refactor) that selects shields for some four hundred networks. There are many more networks to be added, awaiting mostly the artwork. What's there are interstates, Trans-Canada Highway, Mexico Autopistas, state highways for all fifty states, provincial highways of all classes in all ten Canadian provinces and Nunavut, and many county, city and township markers.  (Yes, it's been tested with West Virginia's peculiar double-numbered county roads, and renders them correclty. The route relations aren't available in very many places, but [this map](https://kbk.is-a-geek.net/catskills/test4.html?la=39.5845&lo=-78.0735&z=15) shows a reasonable sampling of state/county secondary/county tertiary hierarchy)

I'm a bit stymied about the path forward from here.  The next steps have to be standing up a tile cache on my server so that I can move away from static rendering, and working on Carto-CSS to make it support `GroupSymbolizer`. The first is something that I'd need to keep my sanity while working on the second, and the second is needed before OSM-Carto would have the functionality available to use.  I'm a little intimidated with the number of big codebases that all of this has an impact on.

Moreover, I freely concded that I'm not a terrific cartographer. About the best that I can say for [the map that I render](https://kbk.is-a-geek.net/catskills/test4.html) is that it works for what I use it for.

**Comment by pnorman** [comment]
#4431 will unblock this

**Comment by pnorman** [comment]
> [#4431](https://github.com/openstreetmap-carto/openstreetmap-carto/pull/4431) already demonstrated an approach suitable for that - to introduce a relations table and use that together with the ways in a JOIN query.

I've since realized that has issues with expiry. You will need to test that a change to a route relation expires the appropriate tiles.

**Comment by pnorman** [comment]
With osm2pgsql tile expiry is the responsibility of the style. See the [osm2pgsql manual](https://osm2pgsql.org/doc/manual.html#expire-outputs) or the [example configs](https://github.com/osm2pgsql-dev/osm2pgsql/blob/master/flex-config/expire.lua). The main issue I see is that flex is designed for multiple tables and the expire outputs are per-table. With only four tables this means many features of the flex backend can't be used as effectively.

Operations might make decisions like cutting off expiry for low zoom tiles because they're re-rendered daily.

For the specific matter of route relations and getting an expire output to reflect the route relation changes, I'm not sure what the best way is. There are two approaches to handling route relations: a JOIN with a route relation table and bringing route relation information onto the linestrings. I've only used the former with styles where they weren't updatable. This wasn't intentional, but it means I don't have a ready-made solution to expiry of tiles where a route relation table is used.
